lib/extended_config.rb

```diff
@@ -0,0 +1,48 @@
+require 'yaml'
+require 'fileutils'
+
+class ExtendedConfig
+  IAAS_TYPE_KEY = 'IAAS_TYPE'.freeze
+  DEFAULT_IAAS_TYPE = 'openstack'.freeze
+
+  attr_reader :extended_config
+  # private_class_method :new
+
+  def default_config
+    {
+      'default' => {
+        'iaas' => default_iaas_type
+      }
+    }
+  end
+
+  def initialize(env = {})
+    @extended_config = env
+  end
+
+  def ==(other)
+    return false unless other.is_a?(ExtendedConfig)
+    @extended_config == other.extended_config
+  end
+
+  private
+
+  def default_iaas_type
+    @extended_config.fetch(IAAS_TYPE_KEY, DEFAULT_IAAS_TYPE)
+  end
+
+  class ExtendedConfigBuilder
```

Comment:
```ruby
def initialize(current_config)
  @current_config = current_config || {}
end
```

Reply:
I don't think that the build method is needed here.

Reply:
I am not sure about the Builder class, I feel that all it does could be done by the ExtendedConfig class without violating the SRP, but ymmv :)

Reply:
At the beginning, I tried to set  `ExtendedConfig.initialize` private to force builder usage, but I didn't succeeded... And trying control properties added to `ExtendedConfig`